numpy/testing/_private/utils.py

```diff
@@ -718,49 +739,32 @@ def chk_same_position(x_id, y_id, hasval='nan'):
                                 names=('x', 'y'), precision=precision)
             raise AssertionError(msg)
 
+        flagged = False
         if isnumber(x) and isnumber(y):
-            has_nan = has_inf = False
             if equal_nan:
-                x_isnan, y_isnan = isnan(x), isnan(y)
-                # Validate that NaNs are in the same place
-                has_nan = any(x_isnan) or any(y_isnan)
-                if has_nan:
-                    chk_same_position(x_isnan, y_isnan, hasval='nan')
+                flagged = func_assert_same_pos(x, y, func=isnan, hasval='nan')
 
             if equal_inf:
-                x_isinf, y_isinf = isinf(x), isinf(y)
-                # Validate that infinite values are in the same place
-                has_inf = any(x_isinf) or any(y_isinf)
-                if has_inf:
-                    # Check +inf and -inf separately, since they are different
-                    chk_same_position(x == +inf, y == +inf, hasval='+inf')
-                    chk_same_position(x == -inf, y == -inf, hasval='-inf')
-
-            if has_nan and has_inf:
-                x = x[~(x_isnan | x_isinf)]
-                y = y[~(y_isnan | y_isinf)]
-            elif has_nan:
-                x = x[~x_isnan]
-                y = y[~y_isnan]
-            elif has_inf:
-                x = x[~x_isinf]
-                y = y[~y_isinf]
-
-            # Only do the comparison if actual values are left
-            if x.size == 0:
-                return
+                flagged |= func_assert_same_pos(x, y,
+                                                func=lambda xy: xy == +inf,
+                                                hasval='+inf')
+                flagged |= func_assert_same_pos(x, y,
+                                                func=lambda xy: xy == -inf,
+                                                hasval='-inf')
 
         elif istime(x) and istime(y):
             # If one is datetime64 and the other timedelta64 there is no point
             if equal_nan and x.dtype.type == y.dtype.type:
-                x_isnat, y_isnat = isnat(x), isnat(y)
+                flagged = func_assert_same_pos(x, y, func=isnat, hasval="NaT")
 
-                if any(x_isnat) or any(y_isnat):
-                    chk_same_position(x_isnat, y_isnat, hasval="NaT")
-
-                if any(x_isnat) or any(y_isnat):
-                    x = x[~x_isnat]
-                    y = y[~y_isnat]
+        if getattr(flagged, 'shape', ()):
```

Comment:
Isn't this attribute guaranteed to exist?

Reply:
No, it can be `False` -- initialized as such. I tried changing, but cannot get rid of the `if flagged` stanza, nor of the `isinstance` - I suggest we leave this to another PR...

Reply:
Just initialize it to `flagged = np.False_`?

Reply:
I thought about that, but I do feel that I'm then decreasing the clarity of the rest of the code - at least in my emacs, `False` is coloured very nicely, and `false_` is not.

Reply:
I'd argue that keeping a variable type consistent is more important than having the color look like in your editor.

> I tried changing, but cannot get rid of the if flagged stanza

I'm gonna do a checkout to see what I'm missing here

Reply:
How about `bool_(False)`, which meets both requirements?

Reply:
Ok, I see why the 0d case still has special handling - I'd forgotten that `arr[bool_0d].shape == (?,) + arr.shape`, and is not 1D.

Reply:
Or an alternate spelling - `np.ndim(flagged) == 0`

Reply:
Yes, I like the `ndim` - that is much clearer why we do it. Pushed up.